File: leetcode/scala/206ReverseLinkedList.scala
// Reverse a singly linked list.

// Example:

// Input: 1->2->3->4->5->NULL
// Output: 5->4->3->2->1->NULL

// Follow up:
// A linked list can be reversed either iteratively or recursively. Could you implement both?
// Definition for singly-linked list.

// Definition for singly-linked list.

class ListNode(var _x: Int = 0) {
  var next: ListNode = null
  var x: Int = _x
}

object Solution {

  def main(args: Array[String]) {
    val node1 = new ListNode(1)
    val node2 = new ListNode(2)
    val node3 = new ListNode(3)
    val node4 = new ListNode(4)
    val node5 = new ListNode(5)
    node1.next = node2
    node2.next = node3
    node3.next = node4
    node4.next = node5
    assert(Print(node1) == "12345")
    val head = reverseList(node1)
    assert(Print(head) == "54321")
  }

  def reverseList(head: ListNode): ListNode = {
    var pre: ListNode = null
    var cur: ListNode = head
    while(cur != null) {
      val tmp = cur.next
      cur.next = pre
      pre = cur
      cur = tmp
    }
    return pre
  }

  def Print(head: ListNode): String = {
    var s: String = ""
    var cur: ListNode = head
    while(cur != null) {
      s += cur.x.toString
      cur = cur.next
    }
    return s
  }
}

File: leetcode/java/206ReverseLinkedList.java
// Reverse a singly linked list.

// Example:

// Input: 1->2->3->4->5->NULL
// Output: 5->4->3->2->1->NULL

// Follow up:
// A linked list can be reversed either iteratively or recursively. Could you implement both?

import java.lang.String;


// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {val = x;}
}


class Solution {
    
    public static void main(String[] args) {
        ListNode node1 = new ListNode(1);
        ListNode node2 = new ListNode(2);
        ListNode node3 = new ListNode(3);
        ListNode node4 = new ListNode(4);
        ListNode node5 = new ListNode(5);
        node1.next = node2;
        node2.next = node3;
        node3.next = node4;
        node4.next = node5;
        assert(toString(node1).equals("12345"));
        ListNode tail = reverseList(node1);
        assert(toString(tail).equals("54321"));
    }


    public static ListNode reverseList(ListNode head) {
        ListNode pre = null;
        ListNode cur = head;
        while(cur != null) {
            ListNode tmp = cur.next;
            cur.next = pre;
            pre = cur;
            cur = tmp;
        }
        return pre;
    }

    public static String toString(ListNode head) {
        String s = "";
        ListNode node = head;
        while (node != null) {
            s = s + String.valueOf(node.val);
            node = node.next;
        }
        return s;
    }
}
